### ISSUE TITLE
release 3.3.0

### ISSUE THREAD

**Issue by jameslamb** [microsoft/LightGBM#4310]
## Summary

I'd like to start a discussion on this question:

> Should we do a v3.3.0 release of LightGBM some time in June?

## Motivation

My primary motivation for this proposal is that user reports in the last two months have revealed critical issues with the R package (#4007, #4045, #4216, #4259, #4305), and these are fixed by recent PRs related to #3016 (especially #4155 and #4247).

But looking at the PRs that have been merged since LightGBM 3.2.1, there are a lot of other useful fixes, including:

* fix for issue in data parallel learning when feature distributions on partitions don't overlap (#4185)
* more accurate handoffs between C++ side and wrapper packages (#4291, #4292)
* changes to MRO to match scikit-learn's preferences (#3192)
* fix warnings from CUDA builds (#4268)

There has only been one technically "breaking" PR merged, and I think it's ok to include in a minor release (#4197).

## Proposal

I'd like to get other maintainers' thoughts on the following proposal.

1. Do not merge #3234 yet
2. Prepare a 3.3.0 release, to be released some time in June
3. Complete any fixes / feature requests that maintainers suggest before that release

Personally, I'd like to include at least the following in a 3.3.0 release:

* C++
    - [x] #4272
    - [x] incorrect calculation of gamma metric in weighted training (#4174 / #4283)
* Dask
    - [x] eval sets (#4101)
    - [ ] *(not critical)* inconsistencies in leaf counts when performing multiclass classification (#4220)
* Python package
    - [x] fix unexplained jupyter kernel restarts (#4301)
* R package

    - [x] prevent segfaults when objects are serialized / deserialized (#4208)
        - [x] https://github.com/microsoft/LightGBM/pull/4586
    - [x] a deprecation warning for `...` in the R package (https://github.com/microsoft/LightGBM/issues/4226#issuecomment-829473584)
        - [x] https://github.com/microsoft/LightGBM/pull/4571
        - [x] https://github.com/microsoft/LightGBM/pull/4581
        - [x] https://github.com/microsoft/LightGBM/pull/4522
        - [x] https://github.com/microsoft/LightGBM/pull/4523
        - [x] https://github.com/microsoft/LightGBM/pull/4548
        - [x] https://github.com/microsoft/LightGBM/pull/4554
        - [x] https://github.com/microsoft/LightGBM/pull/4572
        - [x] https://github.com/microsoft/LightGBM/pull/4573
    - [x] fix memory leaks (#4282)
        - [x] https://github.com/microsoft/LightGBM/pull/4597
    - [x] prevent session crashes on Windows (#4464)
        - https://github.com/microsoft/LightGBM/pull/4496
    - [x] fix `predict()` being broken when using a file (#4034)
        - https://github.com/microsoft/LightGBM/pull/4545
    - [x] updating how autoconf finds libomp on Mac (#4131)
        - https://github.com/microsoft/LightGBM/pull/4506
        - https://github.com/microsoft/LightGBM/pull/4507

I also think this release could be a good opportunity for maintainers to think carefully about what breaking changes might be made (in addition to #3234) in a 4.0.0 release, and to add deprecation warnings for them in the 3.3.0 release.

<hr>

Thanks for your time and consideration!

**Comment by StrikerRUS** [comment]
I'm OK with `3.3.0` release some time in June.

**Comment by shiyu1994** [comment]
June is also OK for me.  I agree that #3234 should not be merged in 3.3.0. And I can help to fix the existing C++ bugs before the release, like #4286.

**Comment by jameslamb** [comment]
Ok, thank you both! I'm currently focused on some of the R items.

**Comment by shiyu1994** [comment]
It seems that we still need some time to finish this release. Is there anything I can help now?

**Comment by StrikerRUS** [comment]
@shiyu1994 Thanks a lot!
I think it will be great if you could take a look at any of the following issues 
- #4156
- #4220
- #4239
- #4286
- #4311
- #4405
- #4414
- #4331

or ones with `[bug]` label and `GPU` word in the title before next release.

**Comment by jameslamb** [comment]
@shiyu1994 apologies, I recently started a new job and haven't had as much time to contribute on LightGBM. I'm going to start working on the R items in https://github.com/microsoft/LightGBM/issues/4310#issue-897541880 in the next few days.

**Comment by shiyu1994** [comment]
Hi @jameslamb Since we are merging the new CUDA version and the new category encoder soon, we need to release 3.3.0 before that. If we cannot fix all the issues in the list in the short term, can we pick some most important ones?

**Comment by jameslamb** [comment]
Yes definitely. Can you please review #4496 (fixes a critical problem affecting the R package on Windows) and provide an opinion on #4543? I think those are the most important remaining items.

**Comment by jameslamb** [comment]
> If we cannot fix all the issues in the list in the short term, can we pick some most important ones?

I just updated the description of this release, marking #4220 and #4282 as `not critical`. I would have preferred getting them fixed before the 3.3.0 release, but I think that it's ok for them to be done later, in the interest of reducing the amount of work left to do to try to get a new minor release out (since it has been a long time since v3.2.1 and many valuable bugfixes and features have not yet been released).

I think I can finish the remaining R package issues in that description by the end of this week. I don't feel strongly about any other documented issues absolutely needing to be done before we do a 3.3.0 release.

**Comment by jameslamb** [comment]
It seems we're almost ready to start the 3.3.0 release!

There are just some PRs that need reviews. @StrikerRUS @jmoralez and I are collaborating on reviewing the remaining R and Python package PRs.

@btrotta @shiyu1994 @guolinke could one of you please help with reviews on #4542 and #4496? These small PRs have been waiting for review for several weeks. @StrikerRUS and I are not confident in our C++ knowledge and are looking for another reviewer to help with those. Thank you very much!

**Comment by shiyu1994** [comment]
Sure. Will do it soon! Sorry for the slow response. I was busy with the new CUDA version in the last a few weeks.

**Comment by shiyu1994** [comment]
It seems that we've done all the items in the checklist (except a _not critical_ one). Congratulations and thank @jameslamb @StrikerRUS for your great effort! Are we now ready for 3.3.0 release?

**Comment by StrikerRUS** [comment]
@shiyu1994 Thanks a lot for your help! Yeah, I think we are ready to start rolling out 3.3.0.

**Comment by jameslamb** [comment]
I agree, I think so too! Please check the maintainer Slack, I left one important question about changes to the R-package's `DESCRIPTION`, but that doesn't have to be resolved to get the process started.

If no one else does, I'll open the release branch PR soon and populate it with the checklist.

**Comment by shiyu1994** [comment]
@jameslamb Sorry for slow response on the Slack. I'll check it frequently from now on.
